picky-krb/src/negoex/data_types.rs

```diff
@@ -0,0 +1,553 @@
+use std::io::{self, Read, Write};
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use num_derive::{FromPrimitive, ToPrimitive};
+use num_traits::{FromPrimitive, ToPrimitive};
+use uuid::Uuid;
+
+use super::{NegoexDataType, CHECKSUM_SCHEME_RFC3961, NEGOEXTS_MESSAGE_SIGNATURE};
+
+const GUID_SIZE: usize = 16;
+pub(crate) const CHECKSUM_HEADER_LEN: u32 = 4 /* header_len */ + 4 /* checksum_scheme */ + 4 /* type */ + 8 /* checksum vector header */;
+
+pub type Guid = Uuid;
+
+impl NegoexDataType for Guid {
+    type Error = io::Error;
+
+    fn size(&self) -> usize {
+        GUID_SIZE
+    }
+
+    fn decode(mut from: impl Read, _message: &[u8]) -> Result<Self, Self::Error> {
+        let mut id_bytes = [0; GUID_SIZE];
+        from.read_exact(&mut id_bytes)?;
+
+        Ok(Self::from_bytes_le(id_bytes))
+    }
+
+    fn encode_with_payload(&self, _offset: usize, mut to: impl Write, _data: impl Write) -> Result<usize, Self::Error> {
+        to.write_all(&self.to_bytes_le())?;
+
+        Ok(0)
+    }
+
+    fn encode(&self, to: impl Write) -> Result<(), Self::Error> {
+        self.encode_with_payload(0, to, &mut [] as &mut [u8])?;
+
+        Ok(())
+    }
+}
+
+/// [2.2.2 GUID typedefs](https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NEGOEX/%5bMS-NEGOEX%5d.pdf)
+/// ```not_rust
+/// typedef GUID CONVERSATION_ID;
+/// ```
+pub type ConversationId = Guid;
+
+/// [2.2.2 GUID typedefs](https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NEGOEX/%5bMS-NEGOEX%5d.pdf)
+/// ```not_rust
+/// typedef GUID AUTH_SCHEME;
+/// ```
+pub type AuthScheme = Guid;
+
+//= message type are always have a size of 4 bytes =//
+const MESSAGE_TYPE_SIZE: usize = 4;
+
+/// [2.2.6.1 MESSAGE_TYPE](https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NEGOEX/%5bMS-NEGOEX%5d.pdf)
+/// ```not_rust
+/// enum
+/// {
+///     MESSAGE_TYPE_INITIATOR_NEGO = 0,
+///     MESSAGE_TYPE_ACCEPTOR_NEGO,
+///     MESSAGE_TYPE_INITIATOR_META_DATA,
+///     MESSAGE_TYPE_ACCEPTOR_META_DATA,
+///     MESSAGE_TYPE_CHALLENGE,
+///     MESSAGE_TYPE_AP_REQUEST,
+///     MESSAGE_TYPE_VERIFY,
+///     MESSAGE_TYPE_ALERT
+/// } MESSAGE_TYPE;
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive)]
```

Comment:
Nitpick. I feel it’s a bit overkill to import and use a proc-macro crate (`num-derive`) for a single struct. Not sure it’s significant in the current state though. Not a blocker.

Reply:
Don't merge for now. Let me fix it. I'm working on it

Reply:
Just pushed a commit with the fix